### PR TITLE
Add Lavamoat to build system

### PR DESCRIPTION
**Pull request by EtDu**
This PR aims to add LavaMoat in the MetaMask build system.

1) Introduces patch-package to work around SES compatibility issues in deep dependencies
2) Update `babel-eslint` to `@babel/eslint-parser` (now in develop)
3) Add explicit imports for babel plugins and presets 
4) Browserify: use explicit require statements to import plugins & transforms
5) Browserify: use explicit imports for transforms in dependencies (required for static analysis)
6) Remove redundant `gulp-sass` assignment to `compiler`
7) Regular config modifications
  - Fix deep endowments bug: remove `.match` from `process.version`
  - Remove `.call` from `hasOwnProperty` & `stream.Stream`
    - Lavamoat fails to dedupe configuration when merging override configuration (E.G. x.y & x.y.z)
8) Override config additions
  - Endow babel presets & plugins manually (dynamic require)
  - Endow <root> to `@babel/core` for reading babel config
  - Endow Browerify transform `loose-envify`
  - Endow `post-css` iterations to `stylelint` (`lazy-import`)
  - incomplete parsing: vinyl, colorette, node-sass 

CLOSED in favor of #9939

### CODE REVIEW

File: patches/error+7.0.2.patch [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+@@ -21,8 +21,11 @@ function TypedError(args) {
+         var errorName = camelCase(args.type) + 'Error';
+         args.name = errorName[0].toUpperCase() + errorName.substr(1);
+     }
++    //extend runs in strict mode under lavamoat, causing failed assignment to .name
++    const modifiedArgs = Object.assign({}, args)
++    delete modifiedArgs.name
+     
+-    extend(createError, args);
++    extend(createError, modifiedArgs);
```
Comment (kriskowal) [inline patches/error+7.0.2.patch:14]: Looks like this was patched in `errors@7.2.0`, so you should be able to surgically upgrade this dependency.

Comment (kriskowal) [reply patches/error+7.0.2.patch:14]: (🙌 @raynos)

Comment (EtDu) [reply patches/error+7.0.2.patch]: Cool! updated

Comment (Raynos) [reply patches/error+7.0.2.patch]: @kriskowal `error@7` is a blast from the past >_>
I rewrote the entire thing to make it less dynamically magical.

**Pull request closed without merging**